### PR TITLE
Caching process update

### PR DESCRIPTION
**Pull request by Anton-Latukha**
* New code included & shares all source code that may be needed.

* New config includes & shares all compiled deps that may be needed in workflows.

* New config stores the source code separately, which means it gets stored one time.

Before this cache stored per Platform/GHC - the sources & compiled results
together. It is while sources not depend on platform or GHC, as sources are
platform & GHC agnostic. Because they were always bundled together
& sources took ~1/2-2/3 of that storage - cache stored multiple duplicates of the same
source code.

`caching` workflow now downloads, builds & shares all sources & all dependencies for all targets.

`dist-newstyle` so far is not included in to the cache, (what to cache from them can be looked into in the next updates).

---

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2467"></a>

### CODE REVIEW

**Comment by Anton-Latukha** [comment]
[skip circleci]

**Comment by Anton-Latukha** [comment]
This PR has WIP commits that allow development & review. Please, it needs to be reviewed as a draft.

File: .github/workflows/bench.yml [hunk 1]
```diff
@@ -47,22 +47,53 @@ jobs:
...
+        sed -i.bak -e 's/haskell-language-server/hls/g' \
+                    -e 's/haskell_language_server/hls/g' \
+                    haskell-language-server.cabal cabal.project
+        sed -i.bak -e 's/Paths_haskell_language_server/Paths_hls/g' \
+                    src/**/*.hs exe/*.hs
+
+    - name: Retrieving `cabal.project` Hackage timestamp
+      run: |
+        # Form: index-state: 2021-11-29T08:11:08Z
+        INDEX_STATE_ENTRY=$(grep index-state cabal.project)
```
Comment (jneira) [inline .github/workflows/bench.yml:72]: nice change, i wanted to do it since some time ago

**Comment by Anton-Latukha** [comment]
Ok. The working results in the PR are in: https://github.com/haskell/haskell-language-server/pull/2467/checks?sha=f97b4bd8794ac9e6ac4121b7cfed8ca92cce08c0

Now cleaning-up the PR from PR-work commits.

**Comment by Anton-Latukha** [comment]
Is there anything I should do to merge?

**Review by jneira — APPROVED** [review]
Looks great overall, separating sources from compiled code is really a good idea

**Comment by Anton-Latukha** [comment]
Nice. After merge, it works as expected. Caching in the `master` started successfully. Happy about this.

**Pull request merged**